pymarsys/connections.py

```diff
@@ -0,0 +1,159 @@
+from abc import ABC, abstractmethod
+import base64
+import datetime
+import hashlib
+import json
+import logging
+import os
+import uuid
+
+import requests
+from aiohttp import ClientSession
+
+EMARSYS_URI = 'https://api.emarsys.net/'
+
+logger = logging.getLogger(__name__)
+
+
+class ApiCallError(Exception):
+    pass
+
+
+class BaseConnection(ABC):
+    """
+    Any connection used to instantiate an Emarsys object or an object inherited
+     from BaseEndpoint should inherit from this class.
+    this class.
+    """
+    @abstractmethod
+    def __init__(self, username, secret, uri):
+        self.username = username
+        self.secret = secret
+        self.uri = uri
+
+    def build_authentication_variables(self):
+        """
+        Build the authentication variables Emarsys' authentication system
+        asks for.
+        :return: nonce, created, password_digest.
+        """
+        nonce = uuid.uuid4().hex
+        created = datetime.datetime.utcnow().strftime(
+            '%Y-%m-%dT%H:%M:%S+00:00'
+        )
+        sha1 = hashlib.sha1(
+            str.encode(nonce + created + self.secret)
+        ).hexdigest()
+        password_digest = bytes.decode(base64.b64encode(str.encode(sha1)))
+        return nonce, created, password_digest
+
+    def build_headers(self, other_headers=None):
+        """
+        Build the headers Emarsys' authentication system asks for.
+        :return: headers.
+        """
+        if not other_headers:
+            other_headers = {}
+        nonce, created, password_digest = \
+            self.build_authentication_variables()
+
+        http_header = ','.join(
+            (
+                'UsernameToken Username="{}"'.format(self.username),
+                'PasswordDigest="{}"'.format(password_digest),
+                'Nonce="{}"'.format(nonce),
+                'Created="{}"'.format(created),
+            )
+        )
+        headers = {
```

Comment:
thoses are http_headers too. and since the include the other ones, they should be called http_header, and the other called wsse_headers probably.